CFI: Strip auto traits off Virtual calls

maurer opened the pull request:
We already use `Instance` at declaration sites when available to glean additional information about possible abstractions of the type in use. This does the same when possible at callsites as well.

The primary purpose of this change is to allow CFI to alter how it generates type information for indirect calls through `Virtual` instances.

This is needed for the "separate machinery" version of my approach to the vtable issues (#122573), because we need to respond differently to a `Virtual` call to the same type as a non-virtual call, specifically [stripping auto traits off the receiver's `Self`](https://github.com/rust-lang/rust/pull/122573/commits/54b15b0c36d4638167732a0542ed0e34ecf17d7c) because there isn't a separate vtable for `Foo` vs `Foo + Send`.

This would also make a more general underlying mechanism that could be used by rcvalle's [proposed drop detection / encoding](https://github.com/rust-lang/rust/pull/116404/commits/edcd1e20a1a69a8590d8ca47b31634854a40e3fb) if we end up using his approach, as we could condition out on the `def_id` in the CFI code rather than requiring the generating code to explicitly note whether it was calling drop.

maurer committed:
CFI: Strip auto traits off Self for virtual calls

Additional trait bounds beyond the principal trait and its implications
are not possible in the vtable. This means that if a receiver is
`&dyn Foo + Send`, the function will only be expecting `&dyn Foo`.

This strips those auto traits off before CFI encoding.

maurer committed:
CFI: Use Instance at callsites

We already use `Instance` at declaration sites when available to glean
additional information about possible abstractions of the type in use.
This does the same when possible at callsites as well.

The primary purpose of this change is to allow CFI to alter how it
generates type information for indirect calls through `Virtual`
instances.

maurer commented:
r? workingjubilee

workingjubilee commented:
@bjorn3 The code says cg_clif uses `Instance` already, yes?

workingjubilee commented on compiler/rustc_codegen_ssa/src/mir/block.rs:
```diff
@@ -147,6 +147,10 @@ impl<'a, 'tcx> TerminatorCodegenHelper<'tcx> {
 
     /// Call `fn_ptr` of `fn_abi` with the arguments `llargs`, the optional
     /// return destination `destination` and the unwind action `unwind`.
+    ///
+    /// You must provide `instance` if it might affect the alias set of the target.
+    /// This will never be the case for direct calls. It is currently only required
+    /// for `Virtual` calls, but providing it when available will avoid complications.
```
I'm somewhat concerned about this requiring such "high-context" information. Most people have no idea what the hell this means by "alias set". Is there truly no other way to get the relevant Instance?

maurer replied:
The only value we have that is a per-instance value in this function right now is the `fn_ptr`, and I don't think the backends generally store instances for those because not all of them have one.

I do think your point about it being high-context whether an instance is necessary is reasonable though. Perhaps it would be better to describe it in terms of things you don't need an `Instance` for?  The reworded comment might read:

You must provide instance unless you are making a non-`Virtual` direct call or calling a raw `fn` pointer.

workingjubilee replied:
sure.

bors commented:
:umbrella: The latest upstream changes (presumably #122580) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented:
given https://github.com/rust-lang/rust/pull/122580 adds an `Instance` arg anyways I withdraw my remarks about that concern and yeah, basically this looks good to me.

workingjubilee commented:
cc @saethlin the reason the Instance is an Option is because no monomorphization may actually have been done, right? and it has to be forwarded all the time now?

saethlin commented:
> the reason the Instance is an Option is because no monomorphization may actually have been done, right?

Nope. It's an `Option` because nobody pressed me on this before :see_no_evil: and I didn't really look into it deeply.

At the bottom of `codegen_call_terminator`, we can have either an `Instance` or just a backend function pointer. So I thought oh well in some cases we don't have an `Instance` so it has to be optional. But upon re-reading the code, I think we do always discover an `Instance` at some point in this function which is the function we're calling. The backend function pointer is derived from this code, which requires that we have found an `Instance` before:
https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/compiler/rustc_codegen_ssa/src/mir/block.rs#L997-L1000

workingjubilee commented:
@saethlin Then it would be correct to simply force all callers to pass `Instance` without leaving them any option about it?

maurer commented:
> > the reason the Instance is an Option is because no monomorphization may actually have been done, right?
> 
> Nope. It's an `Option` because nobody pressed me on this before 🙈 and I didn't really look into it deeply.
> 
> At the bottom of `codegen_call_terminator`, we can have either an `Instance` or just a backend function pointer. So I thought oh well in some cases we don't have an `Instance` so it has to be optional. But upon re-reading the code, I think we do always discover an `Instance` at some point in this function which is the function we're calling. The backend function pointer is derived from this code, which requires that we have found an `Instance` before:
> 
> https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/compiler/rustc_codegen_ssa/src/mir/block.rs#L997-L1000

I belive that the [function pointer](https://github.com/rust-lang/rust/blob/b3df0d7e5ef5f7dbeeca3fb289c65680ad013f87/compiler/rustc_codegen_ssa/src/mir/block.rs#L825) case (non-virtual indirect clal) still doesn't have an instance. Attempting to unwrap the instance at the end also fails empirically.

saethlin commented:
Ah, yes, I see. I didn't trace `def` far enough back up through the function. Bummer, looks like this needs to stay optional.

workingjubilee commented:
alas.

workingjubilee commented:
@maurer This needs to land a test to prevent it from randomly regressing, then.

maurer commented:
I added my usage of this onto this PR to make it testable. I changed the name of the PR to reflect that.

In addition to feeding the `Instance` through, it now uses it in the mangling to remove auto traits from virtual calls. This allows `&(dyn Foo + Send)` receivers to successfully access `&self` functions on the `Foo` trait.

workingjubilee commented:
Nice!

@bors r+

bors commented:
:pushpin: Commit 042e05b1a538bae6a4e91e17460967e867791662 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 042e05b1a538bae6a4e91e17460967e867791662 -->
<!-- homu: {"type":"Approved","sha":"042e05b1a538bae6a4e91e17460967e867791662","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #122582) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

maurer commented:
Conflict resolved, there was a minor `use` statement ordering conflict.

maurer commented:
@rustbot ready

workingjubilee commented:
@bors r+

bors commented:
:pushpin: Commit f434c27067085d72c7770da3b6d4e0bc316fd267 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee f434c27067085d72c7770da3b6d4e0bc316fd267 -->
<!-- homu: {"type":"Approved","sha":"f434c27067085d72c7770da3b6d4e0bc316fd267","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
